### PR TITLE
build: Do not define `PROVIDE_FUZZ_MAIN_FUNCTION` macro unconditionally

### PR DESCRIPTION
**Pull request by hebasto**
No need to define the `PROVIDE_FUZZ_MAIN_FUNCTION` macro when the build system has been configured with the `--disable-fuzz-binary` option.

See https://github.com/bitcoin/bitcoin/pull/24336#pullrequestreview-881368272.

### CODE REVIEW

**Comment by hebasto** [comment]
@MarcoFalke 

Do you think that the fuzz task timeout is caused by this PR change?

**Comment by MarcoFalke** [comment]
No

**Comment by fanquake** [comment]
This should be combined with the `CHECK_RUNTIME_LIB` check rather than another `enable_fuzz_binary` test.

**Comment by MarcoFalke** [comment]
A third refacotring alternative might be to remove it. Then, call `AX_CHECK_LINK_FLAG` unconditionally?

**Comment by hebasto** [comment]
Updated 4a1e159978792bf3edbc1c51e4b9ccdd4c164a7d -> 7e7ae428319f4813087cde1a5f4611ea20bebe97 ([pr24337.01](https://github.com/hebasto/bitcoin/commits/pr24337.01) -> [pr24337.02](https://github.com/hebasto/bitcoin/commits/pr24337.02)).


> This should be combined with the `CHECK_RUNTIME_LIB` check rather than another `enable_fuzz_binary` test.

> A third refacotring alternative might be to remove it. Then, call `AX_CHECK_LINK_FLAG` unconditionally?

Chosen the former suggestion as it assumes minimal diff.

**Comment by hebasto** [comment]
Updated 7e7ae428319f4813087cde1a5f4611ea20bebe97 -> 8d0683da024e62999d574371e965bc8fc51a6f38 ([pr24337.02](https://github.com/hebasto/bitcoin/commits/pr24337.02) -> [pr24337.03](https://github.com/hebasto/bitcoin/commits/pr24337.03)): reverted back and rebased.

> This should be combined with the `CHECK_RUNTIME_LIB` check rather than another `enable_fuzz_binary` test.

Doesn't work. Fails with "multiple definition of `main'" error.

> A third refacotring alternative might be to remove it. Then, call `AX_CHECK_LINK_FLAG` unconditionally?

Doesn't work. The `-DPROVIDE_FUZZ_MAIN_FUNCTION` macro still defined even configured with `--disable-fuzz-binary`.

**Comment by MarcoFalke** [comment]
I mean:

```
if enable_fuzz_binary:
  AX_CHECK_LINK_FLAG
  CHECK_RUNTIME_LIB
fi

**Comment by hebasto** [comment]
> I mean:
> 
> ```
> if enable_fuzz_binary:
>   AX_CHECK_LINK_FLAG
>   CHECK_RUNTIME_LIB
> fi
> ```

Thanks! Done: 8d0683da024e62999d574371e965bc8fc51a6f38 -> 121c4ba4475285dd335655e410aedcb95076ba85 ([pr24337.03](https://github.com/hebasto/bitcoin/commits/pr24337.03) -> [pr24337.04](https://github.com/hebasto/bitcoin/commits/pr24337.04)).

**Comment by hebasto** [comment]
> > I mean:
> > ```
> > if enable_fuzz_binary:
> >   AX_CHECK_LINK_FLAG
> >   CHECK_RUNTIME_LIB
> > fi
> > ```
> 
> Thanks! Done: [8d0683d](https://github.com/bitcoin/bitcoin/commit/8d0683da024e62999d574371e965bc8fc51a6f38) -> [121c4ba](https://github.com/bitcoin/bitcoin/commit/121c4ba4475285dd335655e410aedcb95076ba85) ([pr24337.03](https://github.com/hebasto/bitcoin/commits/pr24337.03) -> [pr24337.04](https://github.com/hebasto/bitcoin/commits/pr24337.04)).

Reverted back.

**Comment by MarcoFalke** [comment]
> failures

That's because it incorrectly uses a previously cached, but unrelated result. You can work around that with:


```diff
diff --git a/configure.ac b/configure.ac
index b7852ae3b5..efdeb186c3 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1300,7 +1300,7 @@ fi
 if test "$enable_fuzz_binary" = "yes"; then
   AC_MSG_CHECKING([whether main function is needed for fuzz binary])
   AX_CHECK_LINK_FLAG(
-    [-fsanitize=$use_sanitizers],
+    [-fsanitize= -fsanitize=$use_sanitizers],
     [AC_MSG_RESULT([no])],
     [AC_MSG_RESULT([yes]); CPPFLAGS="$CPPFLAGS -DPROVIDE_FUZZ_MAIN_FUNCTION"],
     [],

**Comment by MarcoFalke** [comment]
See https://cirrus-ci.com/task/6160465456791552?logs=ci#L2772

> checking whether main function is needed for fuzz binary... checking whether the linker accepts -fsanitize=address,integer,undefined... (**cached**) yes
> no

**Comment by hebasto** [comment]
> fsanitize=address,integer,undefined... (**cached**) yes

Yeap, [`AX_CHECK_LINK_FLAG`](https://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_check_link_flag.m4) creates and uses a cached variable `ax_cv_check_ldflags_$4_$1`.

**Comment by hebasto** [comment]
Updated 8d0683da024e62999d574371e965bc8fc51a6f38 -> c0a089b242150979aefe5238a240ca5de30d4602 ([pr24337.03](https://github.com/hebasto/bitcoin/commits/pr24337.03) -> [pr24337.05](https://github.com/hebasto/bitcoin/commits/pr24337.05)):

 - addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/24337#issuecomment-1046271814)

**Comment by hebasto** [comment]
Rebased c0a089b242150979aefe5238a240ca5de30d4602 -> 3ec7dc7c470a0c117ad601feda5fa7d45e9ecd93 ([pr24337.05](https://github.com/hebasto/bitcoin/commits/pr24337.05) -> [pr24337.06](https://github.com/hebasto/bitcoin/commits/pr24337.06)).

File: configure.ac [hunk 1]
```diff
@@ -1315,21 +1330,8 @@ if test "$enable_fuzz" = "yes"; then
      ]],[[
       */ int not_main() {
      ]])])
-else
-  BITCOIN_QT_INIT
 
-  dnl sets $bitcoin_enable_qt, $bitcoin_enable_qt_test, $bitcoin_enable_qt_dbus
```
Comment (MarcoFalke) [inline configure.ac:1321]: moving this seems unrelated?

Comment (hebasto) [reply configure.ac:1321]: It looks like a way the diff has been rendered. The actual diff is moving `AX_CHECK_LINK_FLAG` stuff.

File: configure.ac [hunk 2]
```diff
@@ -1300,10 +1300,25 @@ if test "$enable_fuzz" = "yes"; then
...
+    QT_DBUS_INCLUDES=SUPPRESS_WARNINGS($QT_DBUS_INCLUDES)
+    QT_TEST_INCLUDES=SUPPRESS_WARNINGS($QT_TEST_INCLUDES)
+  fi
+fi
+
+if test "$enable_fuzz_binary" = "yes"; then
   AC_MSG_CHECKING([whether main function is needed for fuzz binary])
   AX_CHECK_LINK_FLAG(
-    [-fsanitize=$use_sanitizers],
+    [-fsanitize= -fsanitize=$use_sanitizers], dnl Avoid re-use of an unrelated cached result.
```
Comment (fanquake) [inline configure.ac:1321]: This looks like a hack, and the comment doesn't really explain why it's needed. If I understand what it's meant to be doing, couldn't you check that the `AC_LANG_PROGRAM` compiles while passing `SANITIZER_LDFLAGS` as extra flags?

Comment (hebasto) [reply configure.ac:1321]: > This looks like a hack

It is a [hack](https://github.com/bitcoin/bitcoin/pull/24337#issuecomment-1046271814)...

**Comment by hebasto** [comment]
Updated 3ec7dc7c470a0c117ad601feda5fa7d45e9ecd93 -> 1a3c63c0ffbf5ae8b5bc8c72b2457292ffe10e2d ([pr24337.06](https://github.com/hebasto/bitcoin/commits/pr24337.06) -> [pr24337.07](https://github.com/hebasto/bitcoin/commits/pr24337.07), [diff](https://github.com/hebasto/bitcoin/compare/pr24337.06..pr24337.07)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/24337#discussion_r843171705)

File: configure.ac (hunk 2 above)
Comment (hebasto) [reply configure.ac]: [Updated](https://github.com/bitcoin/bitcoin/pull/24337#issuecomment-1089884069).

**Comment by hebasto** [comment]
Updated 3ec7dc7c470a0c117ad601feda5fa7d45e9ecd93 -> c9c4e6cadda85783a005af773a683d44a208bf67 ([pr24337.06](https://github.com/hebasto/bitcoin/commits/pr24337.06) -> [pr24337.07](https://github.com/hebasto/bitcoin/commits/pr24337.07), [diff](https://github.com/hebasto/bitcoin/compare/pr24337.06..pr24337.07)):

- addressed @fanquake's [comment](https://github.com/bitcoin/bitcoin/pull/24337#discussion_r843171705)

**Comment by MarcoFalke** [comment]
Approach ACK c9c4e6cadda85783a005af773a683d44a208bf67 did not review or test 🐤

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Approach ACK c9c4e6cadda85783a005af773a683d44a208bf67 did not review or test 🐤
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiutAv/UkJoJlQodqTvhH4MJ1Hy/QXpFzWdLaVWy3Uts0sXjd+FX7VfiHVaoikh
TuI0Zf22JJuvywfMWfdn1Dk/FTHsk06asCyADFFEzNzZjq5yOkLp6SrCb/CBbulg
TOhOYMYoLuFl/HoJEdUUlwIPmN6Bf26v0vCrrM7/tuoQhZNFBx84wliriyq44n43
FCs5JwcFI84XuDxsI2evH93qLuLuQ50cjw2XocI3cU3MQjleqEOpuVmjeSU2z40m
sBQb7hpCddtJ4PQe/YKYQNA+VAQ9Wz3jS8cVn2PMkhlQTOK+JzkWQEU/H8dC1TxV
dKGcUG9GQoYGVq6UVpP4J0IdfQq5i++1c5uHnxOEH7M+VcUe4LyeAX820S/v4a0n
ORRc5NicHpV7d1TKK4PRdYWL1egWib8dqhmbRuGRS9dvhyk1guhdvxOWQyBPveJi
XiwxUPIIb/N2gWbJsSEIGqMnN4jE17I3o3jXhdwlHR4MGqlX/DaA5zxGBbPRpTSF
aC8Pv1LA
=qSZS
-----END PGP SIGNATURE-----
```


</details>

**Review by fanquake — APPROVED** [review]
ACK c9c4e6cadda85783a005af773a683d44a208bf67 Checked that `PROVIDE_FUZZ_MAIN_FUNCTION` isn't defined when configuring with `--disable-fuzz-binary`.
